dotnet/csharp.bzl

```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
     local = True,
 )
 
-def csharp_configure():
-  """Finds the mono and mcs binaries installed on the local system and sets
-  up an external repository to use the local toolchain.
-
-  To use the local Mono toolchain installed on your system, add the following
-  to your WORKSPACE file:
-
-  ```python
-  csharp_configure()
-  ```
-  """
-  csharp_autoconf(name = "local_config_csharp")
+def _mono_osx_repository_impl(repository_ctx):
+  # TODO(jwall): The below is necessary due to bug:
+  #   https://github.com/bazelbuild/bazel/issues/1235.
+  #   when that bug is fixed the below should be removed.
+  working_dir = repository_ctx.path("")
+  repository_ctx.file(repository_ctx.path("empty"))
+  if not working_dir.exists:
+    fail("working_dir %s does not exist! See https://github.com/bazelbuild/bazel/issues/1235" % working_dir)
+
+  download_output = repository_ctx.path("")
+  #download the package
+  repository_ctx.download_and_extract(
+    repository_ctx.attr._osx_pkg,
+    download_output,
+    repository_ctx.attr._osx_sha256,
+    "", "mono")
+
+  # now we create the build file.
+  toolchain_build = """
+package(default_visibility = ["//visibility:public"])
+exports_files(["mono", "mcs"])
+"""
+  repository_ctx.file("bin/BUILD", toolchain_build)
+
+def _mono_repository_impl(repository_ctx):
+  print("os name: %s" % repository_ctx.os.name)
+  if repository_ctx.attr.use_local:
+    _csharp_autoconf(repository_ctx)
+  elif repository_ctx.os.name == "mac os x":
+    _mono_osx_repository_impl(repository_ctx)
+  else:
+    #if nothing else works look for a local version of mono
+    _csharp_autoconf(repository_ctx)
+
+mono_package = repository_rule(
+  implementation = _mono_repository_impl,
+  attrs = {
+    "_osx_pkg": attr.string(default="http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz"),
```

Comment:
I wasn't clear on the expected style here. In some places I've seen it done this way and in others I've seen it as a constant. I don't have a preference myself. I'll change them to a constant instead.